src/block_service/block_service_manager.cpp

```diff
@@ -33,55 +33,53 @@ block_service_manager::block_service_manager()
-        auto iter = _fs_map.find(provider);
-        if (iter != _fs_map.end())
-            return iter->second.get();
-    }
+    ddebug("close block service manager.");
+    zauto_write_lock l(_fs_lock);
+    _fs_map.clear();
+}
 
+block_filesystem *block_service_manager::get_or_create_block_filesystem(const std::string &provider)
+{
     {
         zauto_write_lock l(_fs_lock);
         auto iter = _fs_map.find(provider);
-        if (iter != _fs_map.end())
+        if (iter != _fs_map.end()) {
             return iter->second.get();
+        }
+    }
 
-        const char *provider_type = dsn_config_get_value_string(
-            (std::string("block_service.") + provider).c_str(), "type", "", "block service type");
+    const char *provider_type = dsn_config_get_value_string(
+        (std::string("block_service.") + provider).c_str(), "type", "", "block service type");
 
-        block_filesystem *fs =
-            utils::factory_store<block_filesystem>::create(provider_type, PROVIDER_TYPE_MAIN);
-        if (fs == nullptr) {
-            derror("acquire block filesystem failed, provider = %s, provider_type = %s",
-                   provider.c_str(),
-                   provider_type);
-            return nullptr;
-        }
+    block_filesystem *fs =
+        utils::factory_store<block_filesystem>::create(provider_type, PROVIDER_TYPE_MAIN);
+    if (fs == nullptr) {
+        derror_f("acquire block filesystem failed, provider = {}, provider_type = {}",
+                 provider,
+                 std::string(provider_type));
+        return nullptr;
+    }
 
-        const char *arguments =
-            dsn_config_get_value_string((std::string("block_service.") + provider).c_str(),
-                                        "args",
-                                        "",
-                                        "args for block_service");
-
-        std::vector<std::string> args;
-        utils::split_args(arguments, args);
-        dsn::error_code err = fs->initialize(args);
-
-        if (dsn::ERR_OK == err) {
-            ddebug("create block filesystem ok for provider(%s)", provider.c_str());
-            _fs_map.emplace(provider, std::unique_ptr<block_filesystem>(fs));
-            return fs;
-        } else {
-            derror("create block file system err(%s) for provider(%s)",
-                   err.to_string(),
-                   provider.c_str());
-            delete fs;
-            return nullptr;
-        }
+    const char *arguments = dsn_config_get_value_string(
+        (std::string("block_service.") + provider).c_str(), "args", "", "args for block_service");
+
+    std::vector<std::string> args;
+    utils::split_args(arguments, args);
+    dsn::error_code err = fs->initialize(args);
+
+    if (dsn::ERR_OK == err) {
+        ddebug_f("create block filesystem ok for provider {}", provider);
+        zauto_write_lock l(_fs_lock);
+        _fs_map.emplace(provider, std::unique_ptr<block_filesystem>(fs));
```

Comment:
Better to check there is no existed block_filesystem with the same name in the map